lib/phoenix_live_view/channel.ex

```diff
@@ -900,16 +946,102 @@ defmodule Phoenix.LiveView.Channel do
-        {:ok, new_socket} -> %{acc | socket: new_socket}
-        {:error, _error_resp, %Socket{} = new_socket} -> %{acc | socket: new_socket}
+      case Upload.put_entries(acc, upload_conf, entries, cid) do
+        {:ok, new_socket} -> new_socket
+        {:error, _error_resp, %Socket{} = new_socket} -> new_socket
       end
     end)
   end
 
-  defp maybe_update_uploads(state, %{} = _payload), do: state
+  defp maybe_update_uploads(%Socket{} = socket, %{} = _payload), do: socket
+
+  defp register_entry_upload(state, from, info) do
+    %{channel_pid: pid, ref: ref, entry_ref: entry_ref, cid: cid} = info
+
+    {new_state, return} =
+      with_socket(state, cid, nil, fn socket, _ ->
+        conf = Upload.get_upload_by_ref!(socket, ref)
+
+        case Upload.register_entry_upload(socket, conf, pid, entry_ref) do
+          {:ok, new_socket, entry} ->
+            reply = %{max_file_size: entry.client_size, chunk_timeout: conf.chunk_timeout}
+            GenServer.reply(from, {:ok, reply})
+            {new_socket, {:ok, nil}}
+
+          {:error, reason} ->
+            GenServer.reply(from, {:error, reason})
+            {socket, :error}
+        end
+      end)
+
+    case return do
+      {:ok, _} -> put_upload_pid(new_state, pid, ref, entry_ref, cid)
+      :error -> {:noreply, new_state}
+    end
+  end
+
+  # If :error is returned, the socket must not change,
+  # otherwise we need to call push_diff on all cases.
+  defp with_socket(state, nil, ref, fun) do
+    {new_socket, return} = fun.(state.socket, nil)
+
+    case return do
+      {:ok, ref_reply} ->
+        {:noreply, new_state} = handle_changed(state, new_socket, ref_reply)
+        {new_state, return}
+
+      {:ok, ref_reply, extra} ->
+        {:noreply, new_state} = handle_changed(state, new_socket, ref_reply)
+        {new_state, extra}
+
+      :error ->
+        {push_noop(state, ref), return}
+    end
+  end
+
+  defp with_socket(state, cid, ref, fun) do
+    %{socket: socket, components: components} = state
+
+    {new_state, diff, return} =
+      case Diff.with_component(socket, cid, %{}, components, fun) do
+        {diff, new_components, return} ->
+          {%{state | components: new_components}, diff, return}
+
+        :error ->
+          {state, :error}
+      end
+
+    case return do
+      {:ok, ref_reply} ->
+        {push_diff(new_state, diff, ref_reply), return}
+
+      {:ok, ref_reply, extra} ->
+        {push_diff(new_state, diff, ref_reply), extra}
+
+      :error ->
+        {push_noop(new_state, ref), return}
+    end
+  end
+
+  defp ensure_unique_upload_name!(state, conf) do
+    upload_ref = conf.ref
+    cid = conf.cid
+    case Map.fetch(state.upload_names, conf.name) do
+      {:ok, {^upload_ref, ^cid}} ->
+        :ok
+
+      :error ->
+        :ok
+
+      {:ok, {_existing_ref, existing_cid}} ->
+        raise RuntimeError,
+              "existing upload for #{conf.name} already allowed in another component #{existing_cid}"
```

Comment:
Can we add more information here? It says what is going wrong but it doesn't tell the user what they can do to fix it, if anything. Or is this an internal LiveView error that should never happen?